File: Coursera/Algorithms_Part_2/Ex44LZWCompression.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.BinaryIn;
import edu.princeton.cs.algs4.BinaryOut;
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.TST;

public class Ex44LZWCompression {

    private static final int R = 256;
    private static final int L = 4096;
    private static final int W = 12;

    public void compress(String input, String output) {
        BinaryIn in = new BinaryIn(input);
        BinaryOut out = new BinaryOut(output);

        TST<Integer> table = new TST<>();
        for (int i = 0; i < R; i++)
            table.put((char) i + "", i);
        int count = R + 1;

        String text = in.readString();
        while (!text.isEmpty()) {
            String s = table.longestPrefixOf(text);
            out.write(table.get(s), W);
            text = text.substring(s.length());
            if (count < L && !text.isEmpty()) table.put(s + text.charAt(0), count++);
        }

        out.write(R, W);
        out.close();
    }

    public void expand(String input, String output) {
        BinaryIn in = new BinaryIn(input);
        BinaryOut out = new BinaryOut(output);

        String[] table = new String[L];
        int count = 0;
        for (; count < R; count++)
            table[count] = (char) count + "";
        table[count++] = "";

        String prev = "";
        while (!in.isEmpty()) {
            int value = in.readInt(W);
            if (value == R) break;
            String s = table[value];
            if (s == null) s = prev + prev.charAt(0);
            out.write(s);
            if (!prev.isEmpty() && count < L)
                table[count++] = prev + s.charAt(0);
            prev = table[value];
        }

        out.close();
    }


    private static void binaryDump(int bitsPerLine, String filename) {

        BinaryIn in = new BinaryIn(filename);

        int count, line = 0;
        StdOut.print(line + ": ");
        for (count = 0; !in.isEmpty(); count++) {
            if (bitsPerLine == 0) {
                in.readBoolean();
                continue;
            }
            else if (count != 0 && count % bitsPerLine == 0) {
                StdOut.println();
                StdOut.print(++line + ": ");
            }
            if (in.readBoolean()) StdOut.print(1);
            else StdOut.print(0);
        }
        if (bitsPerLine != 0) StdOut.println();
        StdOut.println(count + " bits");
    }

    public static void main(String[] args) {

        String inputFile = "mobydick.txt";
        String compressedFile = "compressedFile.txt";
        String expandedFile = "expandedFile.txt";

        Ex44LZWCompression compression = new Ex44LZWCompression();
        compression.compress(inputFile, compressedFile);
        compression.expand(compressedFile, expandedFile);

        binaryDump(200, inputFile);
        binaryDump(200, expandedFile);
        binaryDump(72, compressedFile);

        try {
            StdOut.print("Original: " + (new In(inputFile).readAll()));
        }
        catch (IllegalArgumentException e) {
            StdOut.println("input file is not present");
        }
        try {
            StdOut.println("Output  : " + (new In(compressedFile).readAll()));
        }
        catch (IllegalArgumentException e) {
            StdOut.println("output file is not present");
        }
        try {
            StdOut.print("Expanded: " + (new In(expandedFile).readAll()));
        }
        catch (IllegalArgumentException e) {
            StdOut.println("expanded file is not present");
        }

        try {
            StdOut.println("isSame?" + (new In(inputFile).readAll())
                    .equals(new In(expandedFile).readAll()));
        }
        catch (IllegalArgumentException e) {

        }
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex44LZWCompression.py
from Ex45BinaryIn import BinaryIn
from Ex46BinaryOut import BinaryOut
from Ex36TernarySearchTrieST import TernarySearchTrieST

R = 256
L = 4096
W = 12


def compress(input_name, output_name):

    i_stream = open(input_name)
    o_stream = BinaryOut(output_name)

    table = TernarySearchTrieST()
    for i in range(R):
        table.put(chr(i), i)
    count = R + 1

    text = i_stream.read()
    while len(text) > 0:

        key = table.longest_prefix_of(text)
        o_stream.write_bits_r(table.get(key), W)
        text = text[len(key):]
        if count < L and len(text) > 0:
            table.put(key + text[0], count)
            count += 1

    o_stream.write_bits_r(R, W)
    i_stream.close()
    o_stream.close()


def expand(input_file, output_file):

    i_stream = BinaryIn(input_file)
    o_stream = open(output_file, 'w')

    table = [''] * L
    for i in range(R):
        table[i] = chr(i)
    count = R + 1

    prev = ""
    while not i_stream.is_empty():
        key = i_stream.read_int_r(W)
        if key == R:
            break
        string = table[key]
        if string is None or string == '':
            string = prev + prev[0]
        o_stream.write(string)
        if len(prev) > 0 and count < L:
            table[count] = prev + string[0]
            count += 1
        prev = table[key]

    i_stream.close()
    o_stream.close()


def test_lzw_compression():

    input_file_name = 'input_file.txt'
    compressed_file_name = 'compressed_file.txt'
    output_file_name = 'output_file.txt'

    compress(input_file_name, compressed_file_name)
    expand(compressed_file_name, output_file_name)

    def read_text(file_name):
        try:
            with open(file_name) as f:
                return f.read()
        except FileNotFoundError:
            return ''

    original = read_text(input_file_name)
    compressed = read_text(compressed_file_name)
    output = read_text(output_file_name)

    print()
    print(f'Original   : {original}')
    print(f'Compressed : {compressed}')
    print(f'Output     : {output}')
    print()
    print(f'Is Same? {original == output}')


if __name__ == '__main__':
    test_lzw_compression()
